File: Projects/Software Technologies/ExamPreparation3/Java Skeleton/src/main/java/imdb/controller/FilmController.java
package imdb.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import imdb.bindingModel.FilmBindingModel;
import imdb.entity.Film;
import imdb.repository.FilmRepository;

import java.util.List;

@Controller
public class FilmController {

	private final FilmRepository filmRepository;

	@Autowired
	public FilmController(FilmRepository filmRepository, FilmRepository filmRepository1) {
		//TODO: Implement me ...
		this.filmRepository = filmRepository1;
	}

	@GetMapping("/")
	public String index(Model model) {
		List<Film> films = this.filmRepository.findAll();

		model.addAttribute("view","film/index");
		model.addAttribute("films",films);

		return "base-layout";
	}

	@GetMapping("/create")
	public String create(Model model) {
		model.addAttribute("view","film/create");

		return "base-layout";
	}

	@PostMapping("/create")
	public String createProcess(Model model, FilmBindingModel filmBindingModel) {
		//We get the film's data from the form by using binding model
		//then we save this data to the filmEntity variable.
		Film filmEntity = new Film(
				filmBindingModel.getName(),
				filmBindingModel.getGenre(),
				filmBindingModel.getDirector(),
				filmBindingModel.getYear()
		);

		//We save this film in the databse using filmRepository.
		this.filmRepository.saveAndFlush(filmEntity);
		//Finally we redirect to Home Page.
		return "redirect:/";
	}

	@GetMapping("/edit/{id}")
	public String edit(Model model, @PathVariable int id) {
		//We get the film from the database and save it in the database.
		Film filmEnity = this.filmRepository.findOne(id);

		//We check if it is null
		if (filmEnity == null) {
			return "redirect:/";
		}

		//And if it is not null we render the view with the film
		model.addAttribute("view","film/edit");
		model.addAttribute("film",filmEnity);

		return "base-layout";
	}

	@PostMapping("/edit/{id}")
	public String editProcess(Model model, @PathVariable int id, FilmBindingModel filmBindingModel) {

		//We find the film from the databse by id and store it in filmFromDb variable.
		Film filmFromDb = this.filmRepository.findOne(id);

		//We check if it is null
		if (filmFromDb == null) {
			return "redirect:/";
		}

		//and if it is not null we change the name of the film in the database
		//with the name of the film form the form
		filmFromDb.setName(filmBindingModel.getName());
		//We do this with the genre,
		filmFromDb.setGenre(filmBindingModel.getGenre());
		//And with the director,
		filmFromDb.setDirector(filmBindingModel.getDirector());
		//And with the year too.
		filmFromDb.setYear(filmBindingModel.getYear());

		//Then we save the changes in the database.
		this.filmRepository.saveAndFlush(filmFromDb);

		//And finally we redirect to Home Page.
		return "redirect:/";
	}

	@GetMapping("/delete/{id}")
	public String delete(Model model, @PathVariable int id) {
		//We get the film from the database and save it in the database.
		Film filmEnity = this.filmRepository.findOne(id);

		//We check if it is null
		if (filmEnity == null) {
			return "redirect:/";
		}

		//And if it is not null we render the view with the film
		model.addAttribute("view","film/delete");
		model.addAttribute("film",filmEnity);

		return "base-layout";
	}

	@PostMapping("/delete/{id}")
	public String deleteProcess(@PathVariable int id) {
		//We find the film n the databse
		Film filmFromDb = this.filmRepository.findOne(id);

		//We check if it is not null
		if (filmFromDb == null) {
			return "redirect:/";
		}

		//And if it is not null we delete this film from the database.
		this.filmRepository.delete(filmFromDb);
		return "redirect:/";
	}
}

File: Projects/Software Technologies/ExamPreparation3/C# Skeleton/IMDB/Controllers/FilmController.cs
using System.Linq;
using System.Net;
using System.Web.Mvc;
using IMDB.Models;

namespace IMDB.Controllers
{
    [ValidateInput(false)]
    public class FilmController : Controller
    {
        [HttpGet]
        [Route("")]
        public ActionResult Index()
        {
            using (var db = new IMDBDbContext())
            {
                //We get all movies from the databse.
                var films = db.Films.ToList();

                //And we render the view with the films.
                return View(films);
            }
        }

        [HttpGet]
        [Route("create")]
        public ActionResult Create()
        {
            //We only render the view
            return View();
        }

        [HttpPost]
        [Route("create")]
        [ValidateAntiForgeryToken]
        public ActionResult Create(Film film)
        {
            //With ModelState we check if our form is valid.
            if (ModelState.IsValid)
            {
                //WE create connection wiht the databse
                using (var db = new IMDBDbContext())
                {
                    //We add the film in the databse
                    db.Films.Add(film);
                    //and we save the changes.
                    db.SaveChanges();
                } 
            }
            //finally we redirect to Home Page.
            return RedirectToAction("Index");
        }

        [HttpGet]
        [Route("edit/{id}")]
        public ActionResult Edit(int? id)
        {
            //We check if the id is null because int? id means that this variable can be null.
            if (id == null)
            {
                return HttpNotFound();
            }

            //WE create connection wiht the databse
            using (var database = new IMDBDbContext())
            {
                //We find the film in the databse by id
                Film film = database.Films.First(f => f.Id == id);

                //we check if it is null
                if (film == null)
                {
                    return RedirectToAction("Index");
                }
                //and if it is not we render the edit view with the film.
                return View(film);
            }
            
        }

        [HttpPost]
        [Route("edit/{id}")]
        [ValidateAntiForgeryToken]
        public ActionResult EditConfirm(int? id, Film filmModel)
        {
            //We check if the id is null because int? id means that this variable can be null.
            if (id == null)
            {
                return HttpNotFound();
            }

            //With ModelState we check if our form is valid.
            if (ModelState.IsValid)
            {
                //We create connec tion with the databse.
                using (var database = new IMDBDbContext())
                {
                    //We find the movie thath we want to edit from the database by id.
                    Film filmFromDb = database.Films.Find(id);

                    //We check if it is not null
                    if (filmFromDb == null)
                    {
                        return RedirectToAction("Index");
                    }

                    //And it is not null we coontinue.

                    //This is the actual editing:
                    //We change the name of the film in the database with the the name
                    //of the film from the form
                    filmFromDb.Name = filmModel.Name;
                    //We do this with the genre
                    filmFromDb.Genre = filmModel.Genre;
                    //With the director
                    filmFromDb.Director = filmModel.Director;
                    //and with the year too.
                    filmFromDb.Year = filmModel.Year;

                    database.SaveChanges();
                }
            }
            return RedirectToAction("Index");
        }

        [HttpGet]
        [Route("delete/{id}")]
        public ActionResult Delete(int? id)
        {
            //We check if the id is null because int? id means that this variable can be null.
            if (id == null)
            {
                return HttpNotFound();
            }

            //We create connection with the databse.
            using (var database = new IMDBDbContext())
            {
                //WE get the film from the databse by id
                Film film = database.Films.First(f => f.Id == id);

                //We check if it is null
                if (film == null)
                {
                    return RedirectToAction("Index");
                }

                //and if it is not we render the view with the film in it.
                return View(film);
            }
        }

        [HttpPost]
        [Route("delete/{id}")]
        [ValidateAntiForgeryToken]
        public ActionResult DeleteConfirm(int? id, Film filmModel)
        {
            if (id == null)
            {
                HttpNotFound();
            }

            //We create connection with the databse
            using (var db = new IMDBDbContext())
            {
                //we find the film we want to delete by id.
                var filmToBeDeleted = db.Films.Find(id);

                //we check if it is null
                if (filmToBeDeleted == null)
                {
                    return RedirectToAction("Index");
                }

                //And if it is not we delete it from the database.
                db.Films.Remove(filmToBeDeleted);
                //We save the changes.
                db.SaveChanges();

                // return RedirectToAction("Index");
            }

            return RedirectToAction("Index");
        }
    }
}